Fix libc_nonshared.a

rootbeer opened the pull request:
Build on https://github.com/ziglang/zig/pull/16970 with additional CLs for "weak_hidden_alias" and re-enabling disabled tests.

rootbeer commented:
CI tests are passing.  The two CI failures seem to be `native-native-musl` failures:
> run test std-x86_64-linux-none-Debug-selfhosted-no-lld: error: Illegal instruction at address 0x81febbe
run test std-x86_64-linux-none-Debug-selfhosted-no-lld: error: while executing test 'test.Condition - signal', the following command terminated with signal 4 (expected exited with code 0):

(I'm not clear if there are any subsequent test cases that were skipped because of this failure, or if its isolated to a single test or test file.)

This failure seem unlikely to be due to the changes here because of the "-musl" target .... but they don't look familiar either.  I can rebase or wait for other fixes.  (Or file a new bug.)

Running this stack locally, I can get my "fstat" repro case to pass with a variety of glibc versions (2.19, 2.25, 2.38, etc).  That is an improvement, too.

Oddly, really old glibc versions don't compile (e.g. `-target native-native-gnu.2.10`) --- lots of missing symbols.  I haven't tried them before, so I don't *think* that's related to this CL.  (I'm not sure what the oldest supported glibc should be ... v2.16?  v2.0.0?)

squeek502 commented:
The CI failures are intermittent failures affecting current master as well, so likely unrelated to this PR: https://github.com/ziglang/zig/commits/master

rootbeer commented:
I think this is good to go then.  I'll see if I can add some tests of some sort that cover different glibc versions.

rootbeer commented:
(I don't mean to steal @lifthrasiir's PRs, which are the bulk of this stack, so I'm happy to donate my patches to his stack if that keeps the attributions correct.)

lifthrasiir commented:
> (I don't mean to steal @lifthrasiir's PRs, which are the bulk of this stack, so I'm happy to donate my patches to his stack if that keeps the attributions correct.)

No problem! It seems that my PR is drifted away (again) from the master anyway, so I'll pull this PR after CI passes.

squeek502 commented:
Seems like it'd probably be good to add a standalone test for this. My initial thought was something very simple to test for https://github.com/ziglang/zig/issues/17034 like:

```
const std = @import("std");

test {
    try std.testing.expectError(error.FileNotFound, std.fs.cwd().statFile("test_file_that_doesnt_exist"));
}
```

with different `glibc_version`s in `CrossTarget`s. Unsure how the host's glibc version would affect the usefulness of the test.

Unsure what a test for https://github.com/ziglang/zig/issues/16152 would look like.

rootbeer commented:
I found `test/link/glibc_compat/build.zig`, which I expanded with some more tests.  I compile some simple code (calls fstat, reallocarray, getauxval, and atexit) against a bunch of different glibc versions and makes sure it runs.  I was going to submit it as a separate follow-on PR, but I can add it here if you'd like.  See https://github.com/ziglang/zig/commit/dfde1aeb870870c1d11ff39c5a9432e30a5fd456.

rootbeer commented:
Oh, and my suggested tests are just covering the "positive" cases of glibc symbols being linked correctly.  I don't test things like older versions *not* having newer symbols (e.g., `reallocarray` should not be found in earlier versions).  I could probably figure out how to make a test for this work, if there are some existing tests of linker errors I can follow as an example?

And, I don't test that symbols are left unresolved in the resulting binary (e.g., testing that Zig does not erroneously link in static implementations).    I could probably hack a shell script together using `objdump` or `readelf`, but I'm not sure that's portable enough to be useful?

andrewrk commented:
> I could probably hack a shell script together using `objdump` or `readelf`, but I'm not sure that's portable enough to be useful?

@kubkon might have some build.zig API tips to help with checking objects. For example check out some of the linker tests in test/link/*

rootbeer commented:
I've rebased to today's changes, and I've added two commits for a glibc test case.  

One commit is the basic test case I linked before.  It should help ensure glibc exposes symbols at the correct version during compilation.  The test is executed too, to see that things work at run-time.

I think I figured out how to make the Elf "checkObject()" tests work against the glibc test case, too.  I'm not sure I'm holding the API correctly, but the tests do seem to pass.  And they do seem to fail when I break things.  (I couldn't get it to check that a "FUNC LOCAL HIDDEN" symbol showed up correctly, see the commented out checkExacts.)

rootbeer commented:
Heh, tests fail when running executables linked against the newer glibc versions, as the host can't execute them.  Will fix that.

rootbeer commented:
This stack of commits could use some reviews, I think.  The underlying commits (from @lifthrasiir) to fix libc_nonshared are the bulk of the change, and they're pretty solid and unchanged by me.  In addition to fixing the "fstatat errno problem", this stack removes support for linking glibc versions before v2.17, generally (#17769).  And adds some tests that build against different versions of glibc, and check that dynamic symbols are present as expected.

The tests I added to `test/link/glibc_compat` seem to work, but I'm not sure I'm using the `exe.checkObject()` APIs to check the work of the linker as well I could.

I added a README.md to `lib/libc/glibc/`.  Anyone could review that (especially if you don't know much about Zig's support for glibc).  

The commits that add "bounds checking" to glibc.zig, target.zig, and print_targets.zig seem to work, but I'm not sure if I should use a different approach (e.g., move more of the checks into target.zig), or if the architecture-specific bounds checks are worth the effort.

kubkon commented on test/link/glibc_compat/build.zig:
```diff
@@ -17,4 +27,91 @@ pub fn build(b: *std.Build) void {
+        exe.linkLibC();
+
+        // Only try running the test if the host glibc is known to be good enough
+        if (minor_ver <= running_glibc_minor) {
```
I wonder if this could be incorporated in our runner much like foreign test runner. What do you think @andrewrk ?

rootbeer replied:
It seems complicated to determine the runtime linker's support if there is an emulator in play?  (I guess the test infra could probe for that?)  And I also think these does-runtime-support-newer-glibc checks are probably only needed in this test?  Other tests should just be using the "default" glibc if they're explicitly linking?

That said, I'm willing to try merging this check into the test infra if folks think that is worthwhile.

andrewrk replied:
Use the SemanticVersion methods for comparison please. Also, I think you should be able to remove this condition since you have `skip_foreign_checks = true` which is defined to skip the check if the host cannot execute the binary. If the build system needs an improvement to make this work correctly with respect to glibc versions, I'm happy to help with that.

rootbeer replied:
I will work on making the `skip_foreign_checks = true` check for the glibc supported versions.  Without this hack here, the CI machines would build binaries that failed to execute.  I can do this as a follow-on to this PR, or will update this PR with the changes if you want to wait for that.

(And I've switched to `SemanticVersion.order` for the version checking.)

kubkon commented on test/link/glibc_compat/build.zig:
```diff
@@ -17,4 +27,91 @@ pub fn build(b: *std.Build) void {
+        if (minor_ver < 32) {
+            check.checkExact("0 0 UND FUNC GLOBAL DEFAULT __fxstatat");
+            //check.checkInSymtab();
+            //check.checkExact("{addr1} {size1} {shndx1} FUNC LOCAL HIDDEN fstatat");
```
You should use `{addr}` for extracting values into variables that can later be compared using `checkComputeCompare`. Also, in order to extract values you should use `checkExtract`. Here, I think you wanted to use the same pattern you already used for `.dynsym` check, but a fuzzy check rather than exact match, so something like this:

```
check.checkInSymtab();
check.checkContains("FUNC LOCAL HIDDEN fstatat");
```

rootbeer replied:
Ah, makes sense!   Fixed. 

I've also added some `check.checkNotPresent` tests too ... do those look right?  (I'm not clear if they need their own `checkStart()`?)

kubkon replied:
It depends. If you can guarantee that the next thing checked will be immediately after the previous check, `checkStart` is not needed. However, if that cannot be guaranteed, then there should be a `checkStart` before `checkNotPresent` call (or any other check for that matter).

kubkon reviewed (changes requested):
Tests look good! I just have a few nits I would like you to try.

rootbeer commented on test/link/glibc_compat/build.zig:
```diff
@@ -17,4 +27,91 @@ pub fn build(b: *std.Build) void {
+        // __cxa_atexit.
+        check.checkStart();
+        //check.checkInSymtab();
+        //check.checkExact("{addrX} {sizeX} {shndxX} FUNC LOCAL HIDDEN atexit");
```
Yep.  Fixed.

rootbeer commented on test/link/glibc_compat/glibc_runtime_check.zig:
```diff
@@ -0,0 +1,102 @@
+    var base = std.c.getauxval(std.elf.AT_BASE);
+    var pgsz = std.c.getauxval(std.elf.AT_PAGESZ);
+
+    //std.debug.print("auxval [BASE]={x} [PAGESZ]={x}\n", .{base, pgsz});
```
Removed the stale comments.

rootbeer commented:
Rebased and updated with fixes to the test case.  (I marked the review feedback that I completed as "resolved" in the UI here, hope that's the right protocol.)

andrewrk commented:
Thanks! I've been a bit occupied lately but definitely want to get this landed soon. Thanks for your work.

kubkon reviewed (approved):
Tests look good!

andrewrk commented on lib/libc/glibc/README.md:
```diff
@@ -0,0 +1,85 @@
+`std.os.` namespace is generally what should be used to access C-library
+APIs in Zig code (it is defined depending on the linked C library).
+
+See `src/glibc.zig` for how Zig will build the glibc components.  The core
+C library is sufficient only for compile-time linking (it is a stub
+library that just needs to indicate that a given symbol will be present at
+run-time, it doesn't actually have to implement said symbol).
```
```suggestion
See `src/glibc.zig` for how Zig will build the glibc components.  The generated
shared object files are sufficient only for compile-time linking. They are stub
libraries that only indicate which symbols will be present at
run-time, along with their type and size. The symbols do not contain any actual implementations.
```

rootbeer replied:
Merged this suggestion manually.

andrewrk commented on lib/libc/glibc/README.md:
```diff
@@ -0,0 +1,85 @@
+## Targets
+
+The GNU C Library supports a very wide set of platforms and architectures.
+The current Zig support for glibc only supports Linux.  Extend
+`src/glibc.zig` to support additional platforms.
```
```suggestion
The current Zig support for glibc only supports Linux.
```

rootbeer replied:
Done.

andrewrk commented on lib/libc/glibc/README.md:
```diff
@@ -0,0 +1,85 @@
+currently a handful of patches to these headers to represent new features
+(e.g. `reallocarray`) or changes in implementation (e.g., the `stat()`
+family of functions).
+
```
mention the https://github.com/ziglang/universal-headers/ project. I am likely going to merge https://github.com/ziglang/universal-headers/pull/7 shortly.

edit: you have a mention below that should be moved up here

rootbeer replied:
Moved universal-headers reference up here.

andrewrk commented on lib/libc/glibc/README.md:
```diff
@@ -0,0 +1,85 @@
+# More Information
+
+See
+https://github.com/jacobly0/zig/commit/2314051acaad37dd5630dd7eca08571d620d6496
```
don't link to a third party git repo. I'm pretty sure this commit is in zig's tree as well.

rootbeer replied:
Good eye, not sure how I got this thirdparty URL.  Replaced with the real PR.

andrewrk commented on src/glibc.zig:
```diff
@@ -1119,3 +1138,49 @@ pub fn needsCrtiCrtn(target: std.Target) bool {
+// Glibc v2.17 is the oldest supported linking target due to dependencies
+// from the Zig standard library.
+pub fn isSupportedGlibcVersion(ver: Version) bool {
+    return (ver.major > 2) or ((ver.major == 2) and (ver.minor >= 17));
```
`Version` supports a comparison method already - use it! I see a few other instances of this too.

also this function should just be merged into `isArchSupportedGlibcVersion`

rootbeer replied:
I changed all the version comparisons to use `SemanticVersion.order`.    This function is gone now (more details elsewhere on this review.)

andrewrk commented on src/glibc.zig:
```diff
@@ -1119,3 +1138,49 @@ pub fn needsCrtiCrtn(target: std.Target) bool {
+
+// Check if given glibc version is supported by target's architecture.  Note that
+// target.zig's `available_libcs` should filter out many architectures before this.
+pub fn isArchSupportedGlibcVersion(target: std.Target, ver: Version) !void {
```
this should just return a boolean and the error reporting can happen at the callsite. also it should not be `pub` unless it is called from outside this file.

rootbeer replied:
This function is gone now.

andrewrk commented on src/target.zig:
```diff
@@ -11,7 +11,7 @@ pub const ArchOsAbi = struct {
 
 pub const available_libcs = [_]ArchOsAbi{
-    .{ .arch = .aarch64_be, .os = .linux, .abi = .gnu },
+    .{ .arch = .aarch64_be, .os = .linux, .abi = .gnu }, // glibc aarch64 v2.17+
```
instead of putting this data in a comment, how about putting it into the data? then you can use it in glibc.zig instead of duplicating the data.

rootbeer replied:
Adding the versions here (as an optional `.glibc_min` field) means the existing `canBuildLibC` call in target.zig can do the GNU library version checking too.  And that means the glibc.zig doesn't need to do any version bounds checks, which really simplified things.

andrewrk commented on test/link/glibc_compat/glibc_runtime_check.zig:
```diff
@@ -0,0 +1,95 @@
+// PR #17607 - reallocarray not visible in headers
+inline fn checkReallocarray() !void {
+    // reallocarray was introduced in v2.26
+    if (major > 2 or (major == 2 and minor >= 26)) {
```
use the SemanticVersion method for comparison

rootbeer replied:
Switched to `SemanticVersion.order`.

andrewrk commented on test/link/glibc_compat/glibc_runtime_check.zig:
```diff
@@ -0,0 +1,95 @@
+}
+
+// PR #17607 - reallocarray not visible in headers
+inline fn checkReallocarray() !void {
```
don't mark functions as `inline`

rootbeer replied:
Ah, I thought `inline` was necessary to make the version check happen at comptime.  I've dropped the `inline` and added an explicit `comptime` on the version comparison expression.

andrewrk commented on test/link/glibc_compat/glibc_runtime_check.zig:
```diff
@@ -0,0 +1,95 @@
+}
+
+// getauxval introduced in v2.16
+inline fn checkGetAuxVal() !void {
```
don't mark functions as `inline`

rootbeer replied:
Removed.

andrewrk reviewed (changes requested):
Thanks for your patience, @rootbeer. I finally got around to reviewing this.

Great work. Although I have a lot of review comments, it's almost all surface level stuff. Nice touch with adding those tests.

rootbeer commented on lib/libc/glibc/README.md:
```diff
@@ -0,0 +1,85 @@
+
+By default, Zig binaries will not depend on any external C library, but
+they can be linked against one with the `-lc` option.  The compilation
+target defines which C library: `musl` is the statically linked
```
Ah, I've simplified this description to avoid getting into the static/dynamic distinction here.

rootbeer commented on lib/libc/glibc/README.md:
```diff
@@ -0,0 +1,85 @@
+A specific C library version can be chosen with an appropriate `-target`.
+For example, `-target native-native-gnu.2.19` will use the default CPU and
+OS targets, but will link in a run-time dependency on glibc v2.19 (or
+later).  Use `zig env` to show the default targets and versions.
```
Done

rootbeer reviewed (commented):
Thanks for the thorough review!  I've pushed a new version of this PR that should resolve all but one comment.

I moved the glibc-minimum-version check into the `available_libcs` table in target.zig, and that made the code much simpler and more data-driven (the existing call to `canBuildLibC` now covers the glibc version bounds checking too).  So you might want to look at the changes in src/target.zig since its quite different now.

Also, I haven't yet fixed the `skip_foreign_checks = true` to check for glibc-versions (so there is still some code in build.zig to skip executables that use a newer glibc than the host supports).  I'll look into that next (and either add a commit on this PR, or create a follow-on PR).

rootbeer commented:
I've updated this PR to fix a bug in my glibc bounds checking (it was doing the wrong thing for invalid -- i.e., non-Linux -- targets).

Also, I discovered that `strlcpy` and `strlcat` were new in glibc 2.38, but Zig's glibc header files were not reflecting that.  Added CLs to fix that and test them.  (These are great for testing the "skip_foreign_checks" path because my locally installed v2.36 glibc doesn't have these symbols.)

I spent a bit of time investigating how to extend Run.zig to gracefully hide/handle the case where a test fails because the host's glibc version isn't sufficient for the generated binary.  These dynamic linker failures are fundamentally different from the existing `skip_foreign_checks` handling in Run.zig because the binary actually executes (the dynamic linker returns 1 and prints a message to stderr) where all the existing failures hidden by `skip_foreign_checks` are from errors during exec (the test binary doesn't even start from the parent's point of view).

Decoding the dynamic linker failures after the fact seems sketchy.  I could add preventative code to `runCommand` that preemptively skips these tests if they look like they'll fail, but it seems simpler to keep this check in the glibc_compat/build.zig which is presumably where all the glibc multi-version tests will live anyway?

daurnimator commented on test/link/glibc_compat/glibc_runtime_check.zig:
```diff
@@ -0,0 +1,116 @@
+fn checkStrlcpy_v2_38() !void {
+    var buf: [99]u8 = undefined;
+    const used = c_string.strlcpy(&buf, "strlcpy works!", buf.len);
+    assert(used == 15);
```
This should be
```suggestion
    assert(used == 14);
```

The value does *not* include the nul terminator: https://github.com/bminor/glibc/blob/ef321e23c20eebc6d6fb4044425c00e6df27b05f/string/strlcpy.c#L43

This results in tests failing on systems with glibc >= 2.38.

daurnimator replied:
Sent https://github.com/ziglang/zig/pull/19800 to fix
